File: Cuarto/Programación Distribuida/java_concurrencia3/ejercicio7.java
import java.util.Random;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingQueue;

public class ejercicio7 {
    public static class Data {
        private int puntosDentro;
        private int puntosTotales;

        public Data(int puntosDentro, int puntosTotales) {
            this.puntosDentro = puntosDentro;
            this.puntosTotales = puntosTotales;
        }

        public int getPuntosDentro() {
            return puntosDentro;
        }

        public int getPuntosTotales() {
            return puntosTotales;
        }
    }

    static BlockingQueue<Data> queue = new LinkedBlockingQueue<>(10);

    public static class Consumidor implements Runnable{
        private int puntosDentro;
        private int puntosTotales;
        public Consumidor() {}

        public void run(){
            while(true){
                try {
                    Data data = queue.take();
                    puntosDentro += data.getPuntosDentro();
                    puntosTotales += data.getPuntosTotales();
                    System.out.println("Puntos dentro: " + puntosDentro + " Puntos totales: " + puntosTotales);
                    System.out.println("Pi: " + (4.0 * puntosDentro / puntosTotales) + " Error: " + (Math.PI - (4.0 * puntosDentro / puntosTotales)));
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    public static class Productor implements Runnable{
        private int puntos;
        private int id;
        public Productor(int puntos, int id) {
            this.puntos = puntos;
            this.id = id;
        }

        public void run(){
            
            double x, y;
            int puntosDentro = 0;
            for (int i = 0; i < puntos; i++){
                x = Math.random();
                y = Math.random();
                if (x*x + y*y <= 1) puntosDentro++;
            }
            try {
                Random rand = new Random();
                long espera = rand.nextLong(500, 2000);
                Thread.sleep(espera);
                queue.put(new Data(puntosDentro, puntos));
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            
            System.out.println("Productor " + id + " ha terminado");
            
        }
    }

    public static void main(String[] args) {
        int nThreads = 11;
        Executor exec = Executors.newFixedThreadPool(nThreads);
        exec.execute(new Consumidor());
        for (int i = 0; i < nThreads - 1; i++) {
            exec.execute(new Productor(1000000, i));
        }
    }
}

File: Cuarto/Programación Distribuida/python_concurrencia3/ejercicio7.py
import math
import random
import time
import queue
import concurrent.futures as concurrent

class Productor ():
    def __init__ (self, cola, n, id):
        self.cola = cola
        self.n = n
        self.id = id
    
    def run (self):
        puntos_dentro = 0
        for _ in range(0, self.n):
            x = random.uniform(-1, 1)
            y = random.uniform(-1, 1)
            if x**2 + y**2 <= 1: puntos_dentro += 1
        
        self.cola.put((puntos_dentro, self.n))
        time.sleep(random.uniform(0.5, 2))
        print("Productor ", self.id, " ha terminado")
        return (4*(puntos_dentro/self.n), abs(math.pi - 4*(puntos_dentro/self.n)))

cola = queue.Queue()

executor = concurrent.ThreadPoolExecutor(10)
futures = []
for i in range(0,10):
   futures.append(executor.submit(Productor(cola, 100000, i).run))

for futuro in concurrent.as_completed(futures):
    print("Pi: ", futuro.result()[0], " Error: ", futuro.result()[1])